sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/VersioningTest.groovy

```diff
@@ -0,0 +1,364 @@
+
+        then:
+        receivedValV1 == null
+        valV2 == receivedValV2
+        valV3 == receivedValV3
+    }
+
+    def "List Blobs with Version"() {
+        given:
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def blobItemV2 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+        def blobItemV3 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+
+        when:
+        def blobs = blobContainerClient.listBlobs(new ListBlobsOptions().setDetails(new BlobListDetails().setRetrieveVersions(true)), null)
+
+        then:
+        blobs.size() == 3
+        blobs[0].getVersionId() == blobItemV1.getVersionId()
+        blobs[1].getVersionId() == blobItemV2.getVersionId()
+        blobs[2].getVersionId() == blobItemV3.getVersionId()
+        blobs[0].isCurrentVersion() == null
+        blobs[1].isCurrentVersion() == null
+        blobs[2].isCurrentVersion()
+    }
+
+    def "List Blobs without Version"() {
+        given:
+        blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+        def blobItemV3 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+
+        when:
+        def blobs = blobContainerClient.listBlobs(new ListBlobsOptions().setDetails(new BlobListDetails().setRetrieveVersions(false)), null)
+
+        then:
+        blobs.size() == 1
+        blobs[0].getVersionId() == blobItemV3.getVersionId()
+    }
+
+    def "Begin Copy Blobs with Version"() {
+        given:
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def sourceBlob = blobContainerClient.getBlobClient(generateBlobName())
+        sourceBlob.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+
+        when:
+        def pooler = blobClient.beginCopy(sourceBlob.getBlobUrl(), Duration.ofSeconds(1))
+        def copyInfo = pooler.waitForCompletion().getValue()
+
+        then:
+        copyInfo.getVersionId() != null
+        copyInfo.getVersionId() != blobItemV1.getVersionId()
+    }
+
+    def "Copy From Url Blobs with Version"() {
+        given:
+        blobContainerClient.setAccessPolicy(PublicAccessType.CONTAINER, null)
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def sourceBlob = blobContainerClient.getBlobClient(generateBlobName())
+        sourceBlob.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+
+        when:
+        def response = blobClient.copyFromUrlWithResponse(sourceBlob.getBlobUrl(), null, null, null, null, null, Context.NONE)
+        def versionIdAfterCopy = response.getHeaders().getValue("x-ms-version-id")
```

Comment:
We shouldn't have to index into the raw headers to get this value. Maybe swagger needs to be updated to add this field to the deserialized headers?

Reply:
The swagger document seems to be correct. We do have access to deserialized headers at some point. However, the cases where I am retrieving raw header cover APIs that doesn't seem to have a good place to expose this value - they return Response<Void> or Response<String>, so deserialized headers get lost in the transformation from impl->blob client. I wonder if there's any other option to expose it.
Other alternative I can see is to shoot get properties request after copy but that's not safe in concurrent access scenarios (what if there's other operation that bumps version executed in the middle?).

Reply:
In that case this is fine. I didn't remember that some operations just return a raw Response header

Reply:
We cant call getDeserializedHeaders here?

Reply:
copyFromUrl just returns Response<String>, and Response only has getHeaders, not getDeserializedHeaders